staging/src/k8s.io/api/core/v1/types.go

```diff
@@ -2525,6 +2525,8 @@ type NodeSelector struct {
 type NodeSelectorTerm struct {
 	//Required. A list of node selector requirements. The requirements are ANDed.
 	MatchExpressions []NodeSelectorRequirement `json:"matchExpressions" protobuf:"bytes,1,rep,name=matchExpressions"`
+	//Optional. A list of node selector requirements by node's field. The requirements are ANDed.
```

Comment:
According to the [design doc](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/scheduling/nodeaffinity.md#nodeselector), the NodeSelectorTerm is ANDed; but the implementation and its comments are ORed. Open this issue to confirm the target behaviour. Opened https://github.com/kubernetes/kubernetes/issues/62000 for the discussion :).

Reply:
if I'm reading it correctly, NodeSelector has multiple terms, which are documented to be ORed:

https://github.com/kubernetes/kubernetes/blob/200bc669380528bb5e710608dddb208bc325b2a9/pkg/apis/core/types.go#L2261-L2267

that matches how the code treats that (if any term matches, it returns true, so OR):

https://github.com/kubernetes/kubernetes/blob/200bc669380528bb5e710608dddb208bc325b2a9/pkg/scheduler/algorithm/predicates/predicates.go#L773-L787

each of those terms can have one or more match expressions, which are ANDed together to make a label selector:

https://github.com/kubernetes/kubernetes/blob/200bc669380528bb5e710608dddb208bc325b2a9/pkg/apis/core/types.go#L2269-L2273

and that matches how a label selector is constructed (all match expressions are added as requirements to the selector, so AND):

https://github.com/kubernetes/kubernetes/blob/200bc669380528bb5e710608dddb208bc325b2a9/pkg/apis/core/v1/helper/helpers.go#L215-L247